src/index.ts

```diff
@@ -0,0 +1,149 @@
+enum EmployeStatus {
+  Active = "Active",
+  Inactive = "Inactive",
+  Vacation = "Vacation"
+}
+type DeptBudget = {
+  debit: number
+  credit: number
+}
+type PaymentDetails = {
+  swift: string
+  iBan: string
+  number: number,
+  correspondentBank: string
+}
+type AllEmployeTypes = Employe | PreHiredEmploye
+
+class Company {
+  name: string
+  departaments: Departament[] = []
+  preHiredEmployes: PreHiredEmploye[] = []
+  staff: AllEmployeTypes[] = []
+
+  constructor(name: string) {
+    this.name = name
+  } 
+
+  get allEmployes(): AllEmployeTypes[] {
+    return [...this.departaments.flatMap(({employes}) => employes), ...this.preHiredEmployes]
+  }
+}
+
+class Departament { 
+  static idGenerator: number = 0
+  readonly id: number
+
+  name: string
+  area: string
+  employes: Employe[] = []
+  budget: DeptBudget
+
+  constructor(name: string, area: string, budget: DeptBudget = {debit: 0, credit: 0}) {
+    this.name = name
+    this.area = area
+    this.budget = budget
+    this.id = Departament.idGenerator
+    Departament.idGenerator += 1
+  }
+
+  get balance(): number {
+    return this.budget.debit - this.budget.credit
+  }
+
+  addEmployee(employe: Employe | PreHiredEmploye, paymentDetails?: PaymentDetails): void {
+    if (Utils.isEmploye(employe)) {
+      employe.changeDepartament(this)
+      this.budget.credit += employe.salary
+      this.employes.push(employe)
+    } else if (paymentDetails) {
```

Comment:
This condition concerns me a bit. What if we set that parameter as an empty object?